src/qp/operators/SortMergeJoin.java

```diff
@@ -0,0 +1,451 @@
+                    continue; // stream has closed, skip it.
+                }
+                if (!buffers[i].isEmpty()) {
+                    System.out.println("Buffer[" + i + "] is not empty yet, do not read in new batch from OIS");
+                    continue; // buffer is still loaded
+                }
+                ObjectInputStream run = runs.get(i);
+                System.out.println("Attempt to read in batch for buffer[" + i + "]");
+                try {
+                    Batch data = (Batch) run.readObject();
+                    buffers[i] = data;
+                    //TODO: fill up all buffers. Currently only buffers (index-specific) for non-empty streams are used
+                    System.out.println("Read in batch: ");
+                    Debug.PPrint(buffers[i]);
+                } catch (ClassNotFoundException cnf) {
+                    System.err.println("merge: Class not found for reading batch");
+                    System.exit(1);
+                } catch (EOFException EOF) {
+                    eos[i] = true; // Must remove the stream outside the loop
+                    System.out.println("OIS[" + i + "] has closed");
+                } catch (IOException io) {
+                    System.err.println("merge: Error reading in batch");
+                    System.exit(1);
+                }
+            }
+
+            // TODO: This will change when all buffers are filled up.
+            isMergeComplete = true;
+            for(int i = 0; i < runs.size(); i++) { // only check eos for all runs
+                boolean isOISClosed = eos[i];
+                System.out.println("isOISClosed: " + isOISClosed);
+                isMergeComplete &= isOISClosed;
+            }
+            System.out.println("isMergeComplete: " + isMergeComplete);
+            if (isMergeComplete) {
+                System.out.print("Merge completed: " + mergedRunFileName + "\n\n");
+                continue;
+            }
+
+            // Merge all with existing buffers to fill up output batch
+            while (!out.isFull()) {
+                // Initialise minSoFar
+                Tuple minSoFar = null;
+                Batch chosen = null; // choose arbitrary batch
+
+
+                // TODO: for empty buffers, attempt to read in new batch from its open OIS
+                for (Batch b : buffers) {
+                    if (!b.isEmpty()) {
+                        minSoFar = b.get(0);
+                        chosen = b;
+                        break;
+                    }
+                }
+                if (minSoFar == null) {
+                    // All buffers are empty and output buffer is not full. Write out remaining tuples in output
+                    break;
+                }
+
+                // Iterate through batches and find min tuple of all
+                for (int i = 0; i < buffers.length; i++) {
+                    if (!eos[i] && buffers[i].isEmpty() && i < runs.size()) {
```

Comment:
Can probably just `continue` and read streams into the other buffers. Implementation should be up to us.